packages/lesswrong/server/scripts/importEAGUserInterests.ts

```diff
@@ -33,3 +37,71 @@ Vulcan.importEAGUserInterests = wrapVulcanAsyncScript(
       })
   }
 )
+
+
+const csvSchema = z.object({
+  email: z.string(),
+  careerStage: z.string().optional(),
+  countryOrRegion: z.string().optional(),
+  nearestCity: z.string().optional(),
+  willingnessToRelocateBoston: z.string().optional(),
+  willingnessToRelocateBayArea: z.string().optional(),
+  willingnessToRelocateDC: z.string().optional(),
+  willingnessToRelocateLondon: z.string().optional(),
+  willingnessToRelocateNYC: z.string().optional(),
+  willingnessToRelocateOxford: z.string().optional(),
+  willingnessToRelocateRemote: z.string().optional(),
+  experiencedIn: z.string().optional(),
+  interestedIn: z.string().optional(),
+});
+
+export type UserEAGDetailsRow = z.infer<typeof csvSchema>;
+
+/**
+ * Given a CSV with users' EAG data, import it to the "UserEAGDetails" table, replacing any existing user data.
+ */
+Vulcan.importEAGUserDetails = wrapVulcanAsyncScript(
+  'importEAGUserDetails',
+  async (fileName='EAG_users.csv') => {
+    const now = new Date()
+    const records: Partial<DbUserEAGDetail>[] = []
+    
+    const dataStream: AsyncIterable<AnyBecauseIsInput> = fs.createReadStream(fileName)
+      .pipe(Papa.parse(Papa.NODE_STREAM_INPUT, {header: true, delimiter: ','}))
+
+    for await (const data of dataStream) {
+      const row: UserEAGDetailsRow = csvSchema.parse(data);
+      // do a simple check to see if we have a user with this email address
+      const user = await Users.findOne({email: row.email})
+      if (!user) continue
+
+      records.push({
+        userId: user._id,
+        careerStage: row.careerStage?.split('; '),
+        countryOrRegion: row.countryOrRegion,
+        nearestCity: row.nearestCity,
+        willingnessToRelocate: {
+          Boston: row.willingnessToRelocateBoston,
+          BayArea: row.willingnessToRelocateBayArea,
+          DC: row.willingnessToRelocateDC,
+          London: row.willingnessToRelocateLondon,
+          NYC: row.willingnessToRelocateNYC,
+          Oxford: row.willingnessToRelocateOxford,
+          Remote: row.willingnessToRelocateRemote,
+        },
+        experiencedIn: row.experiencedIn?.split('; '),
+        interestedIn: row.interestedIn?.split('; '),
+        lastUpdated: now
+      })
+    }
+    
+    // upsert EAG data for all the users in the CSV
+    void UserEAGDetails.rawCollection().bulkWrite(records.map((record) => ({
+      updateOne: {
+        filter: {userId: record.userId},
+        update: {$set: record},
+        upsert: true,
```

Comment:
We should probably look at the SQL that this generates to double check the upsert is doing what we expect. Upserts are very janky and (unofficially) deprecated post-migration to Postgres in favour of writing something manually in a repo. I think it's _probably_ fine to leave it for this script (provided we can check the generated SQL is correct), on the understanding that it will probably be broken at some point when we hopefully remove upserts.

Reply:
Oh good to know! Ok I'll add a comment in the code. Ideally this script is only used once, and then if we decide to continue with EAG data we would add some sort of Salesforce integration.